staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch_test.go

```diff
@@ -6713,6 +6713,36 @@ func TestUnknownField(t *testing.T) {
 			ExpectedThreeWay:       `{}`,
 			ExpectedThreeWayResult: `{"array":[1,2,3],"complex":{"nested":true},"name":"foo","scalar":true}`,
 		},
+		"no diff even if null extra": {
+			Original: `{"array":[1,2,3],"complex":{"nested":true},"name":"foo","scalar":true}`,
+			Current:  `{"array":[1,2,3],"complex":{"nested":true},"name":"foo","scalar":true}`,
+			Modified: `{"array":[1,2,3],"complex":{"nested":true},"complex_nullable":{"key":null},"name":"foo","scalar":true}`,
+
+			ExpectedTwoWay:         `{"complex_nullable":{"key":null}}`,
+			ExpectedTwoWayResult:   `{"array":[1,2,3],"complex":{"nested":true},"name":"foo","scalar":true}`,
```

Comment:
why is `complex_nullable` removed here?

Reply:
This is actually the real issue(ref: https://github.com/kubernetes/kubernetes/issues/105689). Because in master branch, after first patch request, `complex_nullable` is preserved but in second patch request(with same commands) it will be removed.

Reply:
Here's what I'm reading in #105689:

Case 1: (working fine)
> If the object's .metadata.labels contains anything other than EFRIED, this works fine: the resulting object has no EFRIED key.
```
Before: {"metadata": {"labels": {"Random": "value"}}}
Patch: {"metadata": {"labels": {"EFRIED": null}}}
After: {"metadata": {"labels": {"Random": "value"}}}
```

Case 2: (broken case, I can't see a test for that?)
> However, if .metadata.labels is initially absent, this command adds it with an EFRIED entry with the empty string as a value
```
Before: {"metadata": {}}
Patch: {"metadata": {"labels": {"EFRIED": null}}}
After: {"metadata": {"labels": {"EFRIED": ""}}}
I would expect (either, I'm not 100% sure, I think both would be fine from OPs perspective): 
{"metadata": {"labels": {}}
OR
{"metadata": {}}
```

Case 3: (working fine?)
> Running the command again removes the entry and the labels.
```
Before: {"metadata": {"labels": {"EFRIED": null}}}
Patch: {"metadata": {"labels": {"EFRIED": null}}}
After: {"metadata": {"labels": {}}
I would expect: {"metadata": {"labels": {}}}
```
I think OP is agreeing with you that `labels` is removed, I think that's a bug. For these situations, I think it should behave exactly like json merge patch does (I'd love to be proven wrong though, I can't seem to verify how json merge patch would behave right now).

Reply:
Case 1 and Case 3 are working as expected. Case 2 is failing.

Case 2 is failing because this [line](https://github.com/kubernetes/kubernetes/blob/876d4e0ab029ac7c314bb0e033bdd036531fe426/staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch.go#L1333l) directly assigns patch value without checking it's inners.

Thus, it causes;

```
Before: {}
Patch: {"complex":{"nested":null}}
After: {"complex":{"nested":null}}
According to the PR correct result: {}
```
or
```
Before: {}
Patch: {"complex":{"nested":null,"nested2":"foo"}}
After: {"complex":{"nested":null,"nested2":"foo"}}
According to the PR correct result: {"complex":{"nested2":"foo"}}
```

(note: I added above tests as unit test in PR)

When I check to assure that is compatible with jsonmerge, I added this test in [here](https://github.com/kubernetes/kubernetes/blob/876d4e0ab029ac7c314bb0e033bdd036531fe426/staging/src/k8s.io/apimachinery/pkg/util/jsonmergepatch/patch_test.go#L229);

```
    original:
      simpleMap:
        key1: 1
    modified:
      simpleMap:
        key1: 1
        key2:
         key3: null
    current:
      simpleMap:
        key1: a
        other: a
    threeWay:
      simpleMap:
        key1: 1
        key2:
          key3: null
    result:
      simpleMap:
        key1: 1
        key2: {}
        other: a
```

It also removes `key3`, the only difference is it added `key2: {}`, but this PR removes also it.

Reply:
Yeah, thanks for that, do you know if we can solve that here? i.e. preserve `key2`.

In that test, we would have:
```
ExpectedTwoWayResult: `{"array":[1,2,3],"complex":{"nested":true},"complex_nullable":{},"name":"foo","scalar":true}`
```

Reply:
@apelisse, thanks for reviews. 

I changed the PR to preserve  `complex_nullable` and also the unit tests according to it.